### PR TITLE
Fhir 16 jf contactpoint component

### PR DESCRIPTION
**Pull request by jmfadeley**
This PR lays out the groundwork for the Contact dropdowns. It hasn't been tweaked yet for many things, but gives an idea of how `CodeableConcepts` could work. Warning! This only covers the FIRST entry for relationship, so you need to figure out how to handle them all!

Steven, you probably need to bring `CenteredTabs`  up in such a way as to be more useful. Maybe have it accept children and adjust the number of tabs accordingly?

### CODE REVIEW

File: components/CenteredTabs.js [hunk 1]
```diff
@@ -1,11 +1,15 @@
+/* eslint-disable react/require-default-props */
+/* eslint-disable react/forbid-prop-types */
```
Comment (orndorffgrant) [inline components/CenteredTabs.js]: I think I've seen this in other places too. If we're disabling these linting rules everywhere we should probably just get rid of them.
If we want to keep them, then we should use the `shape` and `arrayOf` PropTypes instead of `object` and `array`

File: components/CenteredTabs.js [hunk 2]
```diff
@@ -16,20 +20,23 @@ class CenteredTabs extends React.PureComponent {
...
+      value: 1,
     };
   }
 
   handleChange = (event, value) => {
     this.setState({ value });
   };
 
+
+  // This is not reusable. Steven, please take a harder look at this.
```
Comment (orndorffgrant) [inline components/CenteredTabs.js]: Comments:
Two paths I see:

1. make this component only contain the tabs, and a sibling component in the parent will contain the content. In this case, this tabs component would basically pass through props to `Tabs`, and the parent component would control `value`, `onChange`, and the content below the tabs. This is what I had envisioned in my changes to this component in #15 
2. This component contains both the tabs and the content. In which case this component should take a list looking something like
```js
[
{
tabTitle: 'Contacts',
tabComponent: Contacts,
tabProps: { ...contactsProps },
},
...
]
```

In either case, the tab titles should at least be passed as an array prop.

File: components/CenteredTabs.js [hunk 3]
```diff
@@ -41,13 +48,22 @@ class CenteredTabs extends React.PureComponent {
           <Tab label="Overview" />
           <Tab label="Contacts" />
         </Tabs>
-      </>
+        {value === 1 && (
```
Comment (orndorffgrant) [inline components/CenteredTabs.js]: This is probably fine for now, but ideally we won't be comparing `value` to literals, instead comparing to pseudo-enums, or using it as an index in an array of components.

File: components/Contacts.js [hunk 4]
```diff
@@ -1,13 +1,56 @@
...
+
+class Contacts extends PureComponent {
   render() {
+    const { classes, contact } = this.props;
     return (
-      <>
-        {}
-      </>
+      <div className={classes.Contacts}>
+        {contact && contact.map(entry => (
```
Comment (orndorffgrant) [inline components/Contacts.js]: `contact` (singular) is an array?

File: components/Contacts.js [hunk 5]
```diff
@@ -1,13 +1,56 @@
...
+          <InfoExpansion
+            key={`InfoExpansion${uuidv4()}`}
```
Comment (orndorffgrant) [inline components/Contacts.js]: `key`s should be deterministic and consistent for an item. React docs say the list index can be used as a last resort if there isn't already a unique identifier for the item. List index is probably fine here since the list will probably not be reordered.

File: components/Contacts.js [hunk 6]
```diff
@@ -1,13 +1,56 @@
...
+            title={HumanName.preferredName(entry.name) || Organization.preferredDBA(entry.organization)}
```
Comment (orndorffgrant) [inline components/Contacts.js]: I'm not sure how I feel about these helper functions being attached to a component. My inclination is to put them in a `utils` or `helpers` module, but this does organize them nicely too... 🤷‍♂

File: components/HumanName.js [hunk 7]
```diff
@@ -28,49 +31,39 @@ const styles = {
...
-    color: '#243B53',
-  },
-  humanNameTablePeriod: {
-    fontFamily: 'Source Sans Pro',
-    fontSize: '.9em',
-    color: '#829AB1',
-  },
   iconInfo: {
-    color: '#D3D3D3', // Be nice if we could use some preprocessor to render variables real soon.
+    color: '#D3D3D3',
```
Comment (orndorffgrant) [inline components/HumanName.js:52]: should this be one of the theme colors?

File: components/HumanName.js [hunk 8]
```diff
@@ -28,49 +31,39 @@ const styles = {
...
 
 class HumanName extends PureComponent {
+  static preferredName(humanName) {
+    if (humanName === undefined) {
+      return '';
+    }
+    const nameUse = Array.isArray(humanName)
+      ? find(humanName, name => name.use === 'usual') || find(humanName, name => name.use === 'official' || humanName[0])
+      : humanName;
+    return nameUse.text ? nameUse.text : `${nameUse.given.join(' ')} ${nameUse.family}`;
```
Comment (orndorffgrant) [inline components/HumanName.js:65]: These ternaries and `||`s are getting complicated, might be more readable with normal `if` statements

File: components/Organization.js [hunk 9]
```diff
@@ -0,0 +1,38 @@
...
+    if (has(organization, 'alias[0]')) {
+      return organization.alias[0];
+    }
+    return '';
+  }
+
+  render() {
+    const { classes } = this.props;
+    return (
+      <div className={classes.Organization} />
```
Comment (orndorffgrant) [inline components/Organization.js:29]: I'm guessing this will eventually have content but not part of this PR?

File: components/Period.js [hunk 10]
```diff
@@ -0,0 +1,49 @@
...
+    return (`${moment(get(period, 'start')).format('MM/DD/YYYY')
+    } - ${has(period, 'end')
+      ? moment(get(period, 'end')).format('MM/DD/YYYY')
+      : 'Present'}`);
+  }
+
+  render() {
+    const { classes, period, classOverride } = this.props;
+    return (
+      <span className={classOverride || classes.Period}>
```
Comment (orndorffgrant) [inline components/Period.js:36]: This doesn't seem like a scalable way of overriding classes, does MUI not have a nice way of merging prop `classes` with default `classes`?

File: components/shared/DataRow.js [hunk 11]
```diff
@@ -0,0 +1,74 @@
...
+
+const defaultStyles = theme => ({
+  DataRow: {
+    display: 'flex',
+    alignItems: 'baseline',
+    justifyContent: 'space-around',
+    margin: '1em',
+  },
+  DataRowLabel: {
+    fontFamily: 'Source Sans Pro',
```
Comment (orndorffgrant) [inline components/shared/DataRow.js]: should this be in `theme`?

File: components/shared/DataRow.js [hunk 12]
```diff
@@ -0,0 +1,74 @@
...
+    fontSize: '1rem',
+    color: '#486581',
```
Comment (orndorffgrant) [inline components/shared/DataRow.js]: should this be in `theme`?

Comment (orndorffgrant) [reply components/shared/DataRow.js]: theres another color below also hardcoded

File: components/shared/DataRow.js [hunk 13]
```diff
@@ -0,0 +1,74 @@
...
+  },
+  DataRowDetails: {
+    fontFamily: 'Source Sans Pro',
+    fontSize: '.9rem',
+    color: get(theme, 'palette.primary.main', '#829AB1'),
+  },
+});
+
+const DataRow = ({ label, value, details }) => ({
+  render() {
```
Comment (orndorffgrant) [inline components/shared/DataRow.js]: In case this isn't intentional (if it is, disregard): `DataRow` is a function that takes some args and returns a component that will always display those args as its props.
A functional component *is* the render function.
```js
const DataRow = ({ label, value, details, classes }) => (
  <div className={classes.DataRow}>
    <span className={classes.DataRowLabel}>
      {label}
    </span>
    <div className={classes.DataRowGroup}>
      <span className={classes.DataRowValue}>
        {value}
      </span>
      {typeof details === 'object'
        ? details
        : (
          <span className={classes.DataRowDetails}>
            {details}
          </span>
        )
      }
    </div>
  </div>
)
```

File: components/shared/DataRow.js [hunk 14]
```diff
@@ -0,0 +1,74 @@
...
+    return (
+      <div className={classes.DataRow}>
+        <span className={classes.DataRowLabel}>
+          {label}
+        </span>
+        <div className={classes.DataRowGroup}>
+          <span className={classes.DataRowValue}>
+            {value}
+          </span>
+          {typeof details === 'object'
```
Comment (orndorffgrant) [inline components/shared/DataRow.js]: I usually don't like when props could have more than one type

File: util/CodeableConcepts.js [hunk 15]
```diff
@@ -0,0 +1,56 @@
...
+  U: 'Unknown',
+};
+
+export function Relationship(relationship) {
+  if (relationship === undefined || relationship.length === 0) {
+    return '';
+  }
+
+  if (relationship[0].text !== undefined) {
+    return relationship.text;
```
Comment (orndorffgrant) [inline util/CodeableConcepts.js]: if `relationship[0].text !== undefined` then `relationship.text` will definitely be `undefined` right?

**Review by orndorffgrant — COMMENTED** [review]
Just did a code review - looks good! left comments on my stylistic preferences

File: components/CenteredTabs.js (hunk 1 above)
Comment (5t33) [reply components/CenteredTabs.js]: Yeah I think we agreed to not worry about it so I'll remove them

Comment (jmfadeley) [reply components/CenteredTabs.js]: Yeah the difficulty in mapping out these objects is in that they must contain the themes too. So up to Steven if we simply remove/override these rules or consider TypeScript or something.

File: components/CenteredTabs.js (hunk 2 above)
Comment (jmfadeley) [reply components/CenteredTabs.js]: Probably best as another PR but sounds good to me.

File: components/CenteredTabs.js (hunk 3 above)
Comment (jmfadeley) [reply components/CenteredTabs.js]: 🌮 Fair enough.

File: components/Contacts.js (hunk 4 above)
Comment (jmfadeley) [reply components/Contacts.js]: @orndorffgrant https://www.hl7.org/fhir/patient.html yeah, it's not well labeled but it is a zero-to-many cardinality. You can also view the JSON file https://www.hl7.org/fhir/patient.html

File: components/Contacts.js (hunk 5 above)
Comment (jmfadeley) [reply components/Contacts.js]: I went with `uuid` only because of another linting rule that complained about using indexes when mapping keys. I actually agree and prefer indexes, preferably with some kind of prefix involving the component name. Take a gander and feel free to override if need be.

File: components/shared/DataRow.js (hunk 11 above)
Comment (jmfadeley) [reply components/shared/DataRow.js]: If we add the font to theme, yes please.

File: components/shared/DataRow.js (hunk 12 above)
Comment (jmfadeley) [reply components/shared/DataRow.js]: This should be, yes.

File: components/shared/DataRow.js (hunk 13 above)
Comment (jmfadeley) [reply components/shared/DataRow.js]: Yeah this is part of the functional programming approach we prefer over declaring classes.

File: components/shared/DataRow.js (hunk 14 above)
Comment (jmfadeley) [reply components/shared/DataRow.js]: I wasn't all that wild about this approach either. It's versatile but I was left wondering how useful it really was. You can formalize it one way or the other if you prefer.

File: util/CodeableConcepts.js (hunk 15 above)
Comment (jmfadeley) [reply util/CodeableConcepts.js]: It could be tighter. This ONLY checks if it's `undefined` and not empty.

File: components/CenteredTabs.js (hunk 2 above)
Comment (5t33) [reply components/CenteredTabs.js]: Yeah, I planing to make the Contacts a sibling component to the tabs, and use a state variable to determine whether it's visible.

File: components/Contacts.js (hunk 4 above)
Comment (5t33) [reply components/Contacts.js]: I sometimes label FHIR resources like that as plural just for readability sake

File: components/Contacts.js (hunk 5 above)
Comment (5t33) [reply components/Contacts.js]: I'm into indexes, personally. Maybe we can turn off that linting rule.

File: components/HumanName.js (hunk 7 above)
Comment (5t33) [reply components/HumanName.js:52]: Yeah, we're gunna have to change all the colors to the theme colors. I think I'm going to make that one big style update PR.

File: components/HumanName.js (hunk 8 above)
Comment (5t33) [reply components/HumanName.js:65]: Agreed

File: components/shared/DataRow.js (hunk 13 above)
Comment (orndorffgrant) [reply components/shared/DataRow.js]: Right, I'm not sure if I explained what I was saying well: As written, this is not a functional component. A functional component looks like this https://reactjs.org/docs/components-and-props.html#extracting-components

File: components/Organization.js (hunk 9 above)
Comment (5t33) [reply components/Organization.js:29]: Yeah, but we could maybe remove it for now.

File: components/shared/DataRow.js (hunk 11 above)
Comment (5t33) [reply components/shared/DataRow.js]: Yeah, I just added it in my PR. We can update that later in the theme conformance PR with colors/fonts.

File: components/shared/DataRow.js (hunk 13 above)
Comment (5t33) [reply components/shared/DataRow.js]: It seems to me that we can/should be using the typography component here to allow for conformance to a top level theme.

File: components/Organization.js (hunk 9 above)
Comment (orndorffgrant) [reply components/Organization.js:29]: No need to remove if its going to be something later

File: components/shared/DataRow.js (hunk 13 above)
Comment (jmfadeley) [reply components/shared/DataRow.js]: Hm. Sorry @orndorffgrant, I missed your point the first time because I was looking on the phone. Got it. Will write it correctly next time.

File: components/Period.js (hunk 10 above)
Comment (5t33) [reply components/Period.js:36]: I don't think we'll want to be overriding classes. I feel like that's the point of the theme.

File: components/shared/DataRow.js (hunk 13 above)
Comment (5t33) [reply components/shared/DataRow.js]: I didn't realize you could write them that way, to be honest.

Comment (5t33) [reply components/shared/DataRow.js]: It is a pretty common way to do it though.

File: util/CodeableConcepts.js (hunk 15 above)
Comment (5t33) [reply util/CodeableConcepts.js]: relationship.text isn't in line with the fhir spec. I'll update this.

File: components/shared/DataRow.js (hunk 14 above)
Comment (5t33) [reply components/shared/DataRow.js]: I updated this to only be a string. I might make the DataRow a css grid object that takes in 3 component objects instead of strings.
